### PR TITLE
LSM: Block Address Determinism

### PR DESCRIPTION
**Pull request by sentientwaffle**
## Background

To enable efficient recovery and additional verification, TigerBeetle's LSM storage will be deterministic.

There are presently three bugs preventing LSM from being deterministic at checkpoint:

1. All concurrent compactions of a half-measure (from different levels & trees) acquire blocks from the FreeSet concurrently.
2. `SuperBlock.Manifest.append()` is called after `ManifestLog` blocks are written. Manifest block writes from different trees are concurrent with one other, so this is nondeterministic.
3. `ManifestLog.append()` is called concurrently by all running level compactions within tree.

This PR fixes **1** and **2**.
**3** is deferred for a future PR (see below).

## 1: Block Reservation

Previously all concurrent compactions acquire block addresses on an as-needed basis (for both output tables and manifest log blocks).

This is convenient, but because compaction progress is nondeterministic (based on IO ordering) block acquisition order — and therefore storage in general — is likewise nondeterministic.

To fix this:
- Each compaction (per tree/level) reserves an upper bound of block addresses that it may require for compaction.
- The manifest log also reserves the upper bound of block addresses that it may require for compaction.

A compaction or manifest log then acquires a subset of the blocks from its reservation. These acquisitions may be concurrent _between reservations_ without compromising storage determinism. (Within a reservation, acquisitions must still be sequential).

The "lifecycle" of a reservation is documented in `superblock_free_set.zig`.

## 2: `SuperBlock.Manifest.append()`

`SuperBlock.Manifest.append()` is now handled by `ManifestLog.flush()`, which in turn is called synchronously relative to the start of a compaction-beat or checkpoint.

## 3: `ManifestLog.append()`

The `ManifestLog.append()` nondeterminism is _not resolved by this PR._ It will be fixed for free by compaction pacing, assuming we compact the levels of each tree sequentially (with trees still concurrent). (Sequential compaction of levels is also friendlier to SSD garbage collection).

Because this issue isn't fixed yet, the VOPR's `StorageChecker` remains disabled.

## Testing

- `FreeSet` fuzzer:
  - Fuzz the reserve/acquire/forfeit/release lifecycle.
  - Encoding/decoding is covered by the EWAH fuzzer (and extended in an upcoming PR).
- `ManifestLog` fuzzer:
  - Extended to verify `SuperBlock.Manifest.compaction_set` membership.
  - Don't interleave more `append()`s between `compact()`s than the actual Tree will need.
- `Tree`/`Forest` fuzzers: See `8749454`.

## Future Work

In addition to fixing `ManifestLog.append()` determinism, per-beat compaction determinism will enable stricter (smaller) block reservations, which will reduce block fragmentation.

### CODE REVIEW

File: src/vsr/superblock_free_set.zig [hunk 1]
```diff
@@ -96,18 +159,102 @@ pub const FreeSet = struct {
...
+            set.reservation_blocks = 0;
+            set.reservation_session +%= 1;
+            set.reservation_state = .reserving;
+        } else {
+            set.reservation_state = .forfeiting;
+        }
+    }
+
+    /// Marks a free block from the reservation as allocated, and returns the address.
+    /// The reservation must not have been forfeited — that is, they must have been no calls to
```
Comment (jamii) [inline src/vsr/superblock_free_set.zig]: This sentence is incomplete.

File: src/vsr/superblock_free_set.zig [hunk 2]
```diff
@@ -96,18 +159,102 @@ pub const FreeSet = struct {
...
+        assert(set.reservation_state == .reserving);
+        assert(reserve_count > 0);
+
+        var shard_start = find_first_set_bit(
+            set.index,
+            @divFloor(set.reservation_blocks, shard_size),
+            set.index.bit_length,
+        ) orelse return null;
+
+        var block = std.math.max(shard_start * shard_size, set.reservation_blocks);
```
Comment (jamii) [inline src/vsr/superblock_free_set.zig:190]: It wasn't obvious to me at first that Reservation is not a contiguous range but instead promises that there are enough free blocks in that range. That made this code hard to follow. Maybe a comment along the lines of `Find a range which contains reserve_count free blocks`.

Comment (sentientwaffle) [reply src/vsr/superblock_free_set.zig:190]: :+1: I'll make that more clear. There is a note [here](https://github.com/tigerbeetledb/tigerbeetle/blob/5c123869540ad5c45c88171b869cb0a932f20ff3/src/vsr/superblock_free_set.zig#L166-L167) too:

```zig
///   - If a reservation is returned, it covers exactly `reserve_count` free blocks, along with
///     any interleaved already-acquired blocks.
```

**Comment by jamii** [comment]
I cranked up the memory for the forest fuzzer. 

```
> git diff
diff --git a/src/lsm/forest_fuzz.zig b/src/lsm/forest_fuzz.zig
index 39590554..7479075e 100644
--- a/src/lsm/forest_fuzz.zig
+++ b/src/lsm/forest_fuzz.zig
@@ -41,7 +41,7 @@ const Environment = struct {
     const size_max = vsr.Zone.superblock.size().? +
         vsr.Zone.wal_headers.size().? +
         vsr.Zone.wal_prepares.size().? +
-        1024 * 1024 * 1024;
+        16 * 1024 * 1024 * 1024;
```

This seed fails on your branch but not on main:

```
> zig build fuzz_lsm_forest -Drelease-safe=true -- --seed 14709536695519122948
info(fuzz): Fuzz seed = 14709536695519122948
info(lsm_forest_fuzz): fuzz_op_count = 43641
info(lsm_forest_fuzz): fuzz_op_distribution = std.enums.struct:24:19{ .compact = 1.00e+00, .put_account = 8.00e+00, .get_account = 0.00e+00 }
info(lsm_forest_fuzz): puts_since_compact_max = 8191
info(lsm_forest_fuzz): compacts_per_checkpoint = 256
thread 19935 panic: reached unreachable code
/home/jamie/tigerbeetle/src/lsm/groove.zig:0:0: 0x250541 in lsm.groove.GrooveType(test.storage.Storage,tigerbeetle.Account,(struct state_machine.struct:35:14 constant)).compact (fuzz_lsm_forest)
/home/jamie/tigerbeetle/src/lsm/forest.zig:191:59: 0x22872d in Environment.run (fuzz_lsm_forest)
                @field(forest.grooves, field.name).compact(Join.groove_callback(field.name), op);
                                                          ^
/home/jamie/tigerbeetle/src/lsm/forest_fuzz.zig:295:24: 0x224690 in main (fuzz_lsm_forest)
    try Environment.run(&storage, fuzz_ops);
                       ^
/nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/lib/std/start.zig:561:37: 0x210181 in std.start.posixCallMainAndExit (fuzz_lsm_forest)
            const result = root.main() catch |err| {
                                    ^
/nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/lib/std/start.zig:322:5: 0x20fb8e in std.start._start (fuzz_lsm_forest)
    @call(.{ .modifier = .never_inline }, posixCallMainAndExit, .{});
    ^
The following command terminated unexpectedly:
cd /home/jamie/tigerbeetle && /home/jamie/tigerbeetle/zig-cache/o/5f5d2780ae3cfbc3acbba41e15898995/fuzz_lsm_forest --seed 14709536695519122948
error: the following build command failed with exit code 1:
/home/jamie/tigerbeetle/zig-cache/o/18e602c2a5fae2d80096678abf003c6a/build /nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/bin/zig /home/jamie/tigerbeetle /home/jamie/tigerbeetle/zig-cache /home/jamie/.cache/zig fuzz_lsm_forest -Drelease-safe=true -- --seed 14709536695519122948
```

**Comment by jamii** [comment]
It's a confusing crash. It remains if I rebase your branch on master, so it's not something that was recently fixed. But it isn't obviously linked to the reserve changes:

```
thread 20782 panic: reached unreachable code
/nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/lib/std/debug.zig:225:14: 0x24250b in std.debug.assert (fuzz_lsm_forest)
    if (!ok) unreachable; // assertion failure
             ^
/home/jamie/tigerbeetle/src/lsm/tree.zig:541:19: 0x2b3a07 in lsm.tree.TreeType(lsm.table.TableType(u128,lsm.groove.IdTreeValue,lsm.groove.IdTreeValue.compare_keys,lsm.groove.IdTreeValue.key_from_value,554112867134706473364364839029663282043,lsm.groove.IdTreeValue.tombstone,lsm.groove.IdTreeValue.tombstone_from_key),test.storage.Storage,[]const u8{65,99,99,111,117,110,116,46,105,100}).compact (fuzz_lsm_forest)
            assert(op == tree.lookup_snapshot_max);
                  ^
/home/jamie/tigerbeetle/src/lsm/groove.zig:946:31: 0x2a3e13 in lsm.groove.GrooveType(test.storage.Storage,tigerbeetle.Account,(struct state_machine.struct:35:14 constant)).compact (fuzz_lsm_forest)
            groove.ids.compact(Join.tree_callback(.ids), op);
                              ^
/home/jamie/tigerbeetle/src/lsm/forest.zig:191:59: 0x295391 in lsm.forest.ForestType(test.storage.Storage,(struct state_machine.struct:50:49 constant)).compact (fuzz_lsm_forest)
                @field(forest.grooves, field.name).compact(Join.groove_callback(field.name), op);
                                                          ^
/home/jamie/tigerbeetle/src/lsm/forest_fuzz.zig:182:27: 0x294c00 in Environment.compact (fuzz_lsm_forest)
        env.forest.compact(forest_compact_callback, op);
                          ^
/home/jamie/tigerbeetle/src/lsm/forest_fuzz.zig:260:32: 0x2875a8 in Environment.run (fuzz_lsm_forest)
                    env.compact(compact.op);
                               ^
/home/jamie/tigerbeetle/src/lsm/forest_fuzz.zig:295:24: 0x274826 in run_fuzz_ops (fuzz_lsm_forest)
    try Environment.run(&storage, fuzz_ops);
                       ^
/home/jamie/tigerbeetle/src/lsm/forest_fuzz.zig:405:21: 0x26c446 in main (fuzz_lsm_forest)
    try run_fuzz_ops(Storage.Options{
                    ^
/nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/lib/std/start.zig:561:37: 0x26579a in std.start.callMain (fuzz_lsm_forest)
            const result = root.main() catch |err| {
                                    ^
/nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/lib/std/start.zig:495:12: 0x24638e in std.start.callMainWithArgs (fuzz_lsm_forest)
    return @call(.{ .modifier = .always_inline }, callMain, .{});
           ^
/nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/lib/std/start.zig:409:17: 0x245426 in std.start.posixCallMainAndExit (fuzz_lsm_forest)
    std.os.exit(@call(.{ .modifier = .always_inline }, callMainWithArgs, .{ argc, argv, envp }));
                ^
/nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/lib/std/start.zig:322:5: 0x245232 in std.start._start (fuzz_lsm_forest)
    @call(.{ .modifier = .never_inline }, posixCallMainAndExit, .{});
    ^
```

File: src/lsm/forest_fuzz.zig [hunk 3]
```diff
@@ -199,11 +201,12 @@ const Environment = struct {
         env.change_state(.forest_checkpointing, .superblock_checkpointing);
         env.superblock.checkpoint(superblock_checkpoint_callback, &env.superblock_context, .{
             .commit_min_checksum = env.superblock.working.vsr_state.commit_min_checksum + 1,
-            .commit_min = env.superblock.working.vsr_state.commit_min + 1,
```
Comment (jamii) [inline src/lsm/forest_fuzz.zig:202]: Oh, this is suspicious. Maybe the cause of the fuzzer crash?

File: src/lsm/manifest_log.zig [hunk 4]
```diff
@@ -78,29 +77,44 @@ pub fn ManifestLogType(comptime Storage: type, comptime TableInfo: type) type {
...
 
-        const block_body_size = config.block_size - @sizeOf(vsr.Header);
-        const entry_size = @sizeOf(Label) + @sizeOf(TableInfo);
-        const entry_count_max_unaligned = @divFloor(block_body_size, entry_size);
-        pub const entry_count_max = @divFloor(entry_count_max_unaligned, alignment) * alignment;
+        /// The maximum number of table updates to the manifest by a half-measure of table
+        /// compaction (not including manifest log compaction).
+        ///
+        /// Input tables are removed from the manifest.
+        /// Output tables are removed from the manifest.
```
Comment (jamii) [inline src/lsm/manifest_log.zig]: Output tables are inserted into the manifest?

Comment (sentientwaffle) [reply src/lsm/manifest_log.zig]: Thanks, typo.

File: src/lsm/forest_fuzz.zig (hunk 3 above)
Comment (sentientwaffle) [reply src/lsm/forest_fuzz.zig:202]: Yep, should be

```zig
.commit_min = env.checkpoint_op.?,
```

instead of

```zig
.commit_min = env.checkpoint_op.? + 1,
```

File: src/lsm/tree.zig [hunk 5]
```diff
@@ -51,6 +51,21 @@ const half_bar_beat_count = @divExact(config.lsm_batch_multiple, 2);
 /// The maximum number of tables for a single tree.
 pub const table_count_max = table_count_max_for_tree(config.lsm_growth_factor, config.lsm_levels);
 
+/// The upper-bound count of input tables to a single tree's compaction.
+///
+/// +1: from level A
+/// +1: from level B, with one edge overlapping the lsm_growth_factor tables.
+///     (This is not +2 because input table selection is least-overlap. If the input table overlaps
+///     on both edges, there must be another table with less overlap to select).
+pub const compaction_tables_input_max = 1 + config.lsm_growth_factor + 1;
```
Comment (jamii) [inline src/lsm/tree.zig]: I'm still confused about why this is +1 and not +0. 

Suppose every table in level A overlaps at least growth_factor + 1 tables in level B.
Assign each table in level B to the first table in level A that overlaps it.
Then the first table in level A has at least growth_factor + 1 tables assigned to it, and every other table in level A has at least growth_factor tables assigned.
So the total number of tables in level B must be at least (level_a_table_count * growth_factor) + 1, which is too many.

**Comment by jamii** [comment]
Everything else looks fine.

File: src/lsm/tree.zig (hunk 5 above)
Comment (sentientwaffle) [reply src/lsm/tree.zig]: You're correct; I just didn't understand that before :+1:.

File: src/vsr/superblock_free_set.zig (hunk 1 above)
Comment (jamii) [reply src/vsr/superblock_free_set.zig]: Also probably s/they/there

**Review by jamii — APPROVED** [review]
